app/server/lib/AttachmentFileManager.ts

```diff
@@ -0,0 +1,217 @@
+    if (!store) {
+      this._log.info({ fileIdent, storeId }, "tried to fetch attachment from an unavailable store");
+      throw new StoreNotAvailableError(storeId);
+    }
+    return this._addFileToAttachmentStore(store, fileIdent, fileData);
+  }
+
+  public async getFileData(fileIdent: string): Promise<Buffer> {
+    const fileInfo = await this._docStorage.getFileInfo(fileIdent);
+    if (!fileInfo) {
+      this._log.error({ fileIdent }, "cannot find file metadata in document");
+      throw new MissingAttachmentError(fileIdent);
+    }
+    this._log.debug(
+      { fileIdent, storeId: fileInfo.storageId },
+      `fetching attachment from ${fileInfo.storageId ? "external" : "document "} storage`
+    );
+    if (!fileInfo.storageId) {
+      return fileInfo.data;
+    }
+    const store = await this._getStore(fileInfo.storageId);
+    if (!store) {
+      this._log.warn({ fileIdent, storeId: fileInfo.storageId }, `unable to retrieve file, store is unavailable`);
+      throw new StoreNotAvailableError(fileInfo.storageId);
+    }
+    return this._getFileDataFromAttachmentStore(store, fileIdent);
+  }
+
+  private async _addFileToLocalStorage(fileIdent: string, fileData: Buffer): Promise<AddFileResult> {
+    const isNewFile = await this._docStorage.findOrAttachFile(fileIdent, fileData);
+
+    return {
+      fileIdent,
+      isNewFile,
+    };
+  }
+
+  private async _getStore(storeId: AttachmentStoreId): Promise<IAttachmentStore | null> {
+    if (!this._storeProvider) {
+      throw new StoresNotConfiguredError();
+    }
+    return this._storeProvider.getStore(storeId);
+  }
+
+  private _getDocPoolId(): string {
+    if (!this._docPoolId) {
+      throw new StoresNotConfiguredError();
+    }
+    return this._docPoolId;
+  }
+
+  private async _getFileIdentifier(fileExtension: string, fileData: Readable): Promise<string> {
+    const checksum = await checksumFileStream(fileData);
+    return `${checksum}${fileExtension}`;
+  }
+
+  private async _addFileToAttachmentStore(
+    store: IAttachmentStore, fileIdent: string, fileData: Buffer
+  ): Promise<AddFileResult> {
+    const isNewFile = await this._docStorage.findOrAttachFile(fileIdent, undefined, store.id);
+
+    // Verify the file exists in the store. This allows for a second attempt to correct a failed upload.
+    const existsInRemoteStorage = !isNewFile && await store.exists(this._getDocPoolId(), fileIdent);
+
+    if (!isNewFile && existsInRemoteStorage) {
+      return {
+        fileIdent,
+        isNewFile: false,
+      };
+    }
+
+    // Possible issue if this upload fails - we have the file tracked in the document, but not available in the store.
+    // TODO - Decide if we keep an entry in SQLite after an upload error or not. Probably not?
+    await store.upload(this._getDocPoolId(), fileIdent, Readable.from(fileData));
+
+    // TODO - Confirm in doc storage that it's successfully uploaded? Need to decide how to handle a failed upload.
```

Comment:
Since external storage is not entirely under our control (an admin might reasonably delete some files there) we need to degrade gracefully if some attachments are mysteriously missing from expected location. Presumably that means that even without this TODO, a failure is non-catastrophic in the sense that the document will continue to work.

Reply:
It's not well explained in this PR, but the transfer code defines this error handling more explicitly.

In the case that the file isn't accessible in external storage, it should show that the attachment isn't available in the UI.  I think that's handled on the clientside when the API returns an error response?